### DISCUSSION TITLE
warpto

### DISCUSSION THREAD

**Discussion by wepl** [vAmigaWeb/vAmigaWeb#212] [Ideas]
The **warpto** option is a nice idea.
The problem is that the booting times of AROS and CBM ROMs are very different. AROS takes much longer to boot.
Also the timing may depend on CPU type and speed.
I think it would be more handy to have an option to let start the emu in warp mode. And having a cmd which could be run (e.g. from the startup-sequence) to stop the warp mode. With that it would be easier to setup a fast startup and return to normal speed at the right position.
In WinUAE there is uaecontrol for this.

**Comment by mithrendal** [comment]
what about writing a string command to ser: ? 

something like `type "warp off" to ser:`

**Reply by wepl** [reply]
that would be fine for me :)

**Comment by dirkwhoffmann** [comment]
I've opened up issue dirkwhoffmann/vamiga#820 for this request.

**Comment by mithrendal** [comment]
Ok dirk implemented the new Retro-Shell commander device which can be connected to the serial port of the virtual Amiga… 

I will merge the code into vAmigaWeb in the coming days… 

Goal:

1.API to connect retroshell commander in the vAmigaWeb player startup script, so that the retroshell commander is listening for commands send from inside the Amiga via ser:

2.CLI command `echo „amiga set warpmode WARP_NEVER“ to ser:` has to end the **_warp during boot feature_** aka warpto

**Comment by dirkwhoffmann** [comment]
Please take into account that warp boot overrides the selected warp mode. To get the intended behavior, ignore warp boot. Instead, start with warp mode `WARP_ALWAYS` and turn it off later by switching to `WARP_NEVER` with the ser: approach.

**Reply by mithrendal** [reply]
roger houston ground control ... 

<img width="774" alt="image" src="https://github.com/vAmigaWeb/vAmigaWeb/assets/17108995/e3b5fda2-22cb-41ae-ad93-8c6f8e7885c7">


I see ... we will bake that into vAmigaWeb as you propose  ...


oh wait ... vamigaweb does not have an option for WARP_ALWAYS, does anyone mind if we exploit the existing call_param `warpto` frame count ?  e.g. maybe warpto=-1 could translate to WARP_ALWAYS ?

oh maybe still better than warpto=-1 is that we don't use the callparam `warpto` and instead in the vAmigaWeb_player startupscript we call wasm_configure("WARP_MODE", "WARP_ALWAYS") ... yeah I think I we should go that route ...

like in the following player config 

```javascript
let config=
{
	AROS:true,
	navbar:false,
	wide:true,
	startup_script: `
hd_mount_list=['UnitA_FA18Interceptor.hdf'];
wasm_configure('WARP_MODE', 'WARP_ALWAYS')
`
};

```

I will try...

**Comment by dirkwhoffmann** [comment]
> `wasm_configure("WARP_MODE", "WARP_ALWAYS")`

Yes, that's the way to go.

**Comment by mithrendal** [comment]
I will add to new actions for the action buttons too...  so one can create buttons or keyboard shortcuts for these two warp modes

<img width="971" alt="image" src="https://github.com/vAmigaWeb/vAmigaWeb/assets/17108995/d8c6bcf3-bbe5-4dad-970c-fcb714c6a539">

**Comment by mithrendal** [comment]
@wepl published new `vAmigaWeb version 2024.03.07` with with retroshell commander listening to Amigas serial port 😎


on your embedding website in the vAmigaWeb players `startup-script` do the following

1.connect retroshell-commander device to serial port
2.set warp_mode to always 

```javascript
let config=
{
	AROS:true,
	navbar:false,
	wide:true,
	startup_script: `
hd_mount_list=['UnitA_FA18Interceptor.hdf'];
wasm_configure("SER_DEVICE", "COMMANDER");
wasm_configure('WARP_MODE', 'WARP_ALWAYS');
`
};
```


on Amiga side e.g. in startup-sequence, set warp mode to NEVER or AUTO
```
echo „amiga set warpmode WARP_NEVER“ to ser:
```

**Reply by wepl** [reply]
I tried but get a JavaScript error.
https://whdload.de/ctros/UnitA_FA18Interceptor.html
https://vamigaweb.github.io/#%7B%22AROS%22%3Atrue%2C%22navbar%22%3Afalse%2C%22wide%22%3Atrue%2C%22startup_script%22%3A%22%5Cn%5Ct%5Ct%5Ct%5Cthd_mount_list%3D%5B'UnitA_FA18Interceptor.hdf'%5D%3B%5Cn%5Ct%5Ct%5Ct%5Ctwasm_configure('SER_DEVICE'%2C'COMMANDER')%3B%5Cn%5Ct%5Ct%5Ct%5Ctwasm_configure('WARP_MODE'%2C'WARP_ALWAYS')%3B%5Cn%5Ct%5Ct%5Ct%22%7D

**Comment by mithrendal** [comment]
probably it launched with the older version, which did not had the retro commander and therefore the configure produced that exception...


please try to reload... it should bring a small update dialog in the right bottom corner and then check if it persists with todays version too

vAmigaWeb always first tries to load from local app cache,  when there is a new version of vAmigaWeb, it will inform the user about it and it asks him whether upgrade to it or stay on the older version... of course when private/incognito browsing is enabled then it will always load the version from the web server

**Reply by mithrendal** [reply]
Ah wait I see because of the error it can not update because it crashes before the update dialog shows up...

let me investigate ...

**Reply by wepl** [reply]
Thanks, works now.
A pity that AROS doesn't have SER:.
Will look for some tools to handle this.

**Reply by mithrendal** [reply]
the problem with the old vamigaweb version is that it does not know the new configure commands and therefore crashes before it can show the update for the new version...

one way would be to handle this with a try catch
```javascript
hd_mount_list=['UnitA_FA18Interceptor.hdf'];
try{
  wasm_configure('SER_DEVICE','COMMANDER');
  wasm_configure('WARP_MODE','WARP_ALWAYS');
} catch{}
```

when executed with the old version... it would simply skip the configure commands

**Reply by mithrendal** [reply]
I checked the configure commands  in your example , they do not turn the warp on 😬

I think the configure happens too early in startup... and will consequently be overwritten with defaults settings from the vAmigaCore

lets do the configure a bit later when the core sends us the READY_TO_RUN message, I.e. after all initialization already happened and just before we call run on the Core 

```javascript
hd_mount_list=['UnitA_FA18Interceptor.hdf'];
on_ready_to_run= ()=>
{
  wasm_configure('SER_DEVICE','COMMANDER');
  wasm_configure('WARP_MODE','WARP_ALWAYS');
}
```

I just published the version `2024_03_09` which adds support for this new `on_ready_to_run hook`... older versions will ignore the hook (nothing happens) also the try{} catch {} is not necessarily needed this way anymore as the older versions of vAmigaWeb will just ignore the hook

**Comment by wepl** [comment]
How can I see if warp mode is active?
I think the serial approach is complicated.
AROS doesn't have SER: it seems. And SER: on CBM requires serial.device which I cannot provide for copyright reasons.
So I need a 3rd-party tool.
I tried SendSerial + artser.device from aminet. But I don't see any output nor that speed changes.
It would be useful, that when sending something via serial there would be some feedback/acknowledge. Don't know if this is technical possible.

**Reply by mithrendal** [reply]
>How can I see if warp mode is active?
>It would be useful, that when sending something via serial there would be some feedback/acknowledge. 


you need first to turn on the logging console in the settings. when turned on it will write to the javascript console.  

<img width="837" alt="image" src="https://github.com/vAmigaWeb/vAmigaWeb/assets/17108995/f8a21b98-8e29-4984-94a3-70d00c25a3b7">

now the logging console will be loaded into the web app and a little icon appears


<img width="1010" alt="image" src="https://github.com/vAmigaWeb/vAmigaWeb/assets/17108995/614f5d3c-92ca-4fba-9134-962b1ff6cf58">


you can bring up the console with a click on the icon

or

you can open the browsers logging console for output 


lets send some thing to the serial port

<img width="1010" alt="image" src="https://github.com/vAmigaWeb/vAmigaWeb/assets/17108995/0a6a524d-2406-4691-8ba1-06dfb6db7985">

a look into the logging console reveals that it sent the bytes 104, 101, 2x 108, 111, 10.

which corresponds to the ASCII values of the word h e l l o

the last byte 10 ist the newline char


now lets look into how to see whether it is warping.

<img width="1010" alt="image" src="https://github.com/vAmigaWeb/vAmigaWeb/assets/17108995/8c033ed5-d7ad-40b3-8257-919b1337d60d">

I turned warp to always and it shows this like above in the log console 

 
>AROS doesn't have SER: it seems.

is it possible to directly write to the amiga port registers ?  @dirkwhoffmann do you have an idea?

**Reply by wepl** [reply]
Thanks!
Is there a way to reset the config of vAmigaWeb? Delete Browser data?

**Reply by mithrendal** [reply]
>Is there a way to reset the config of vAmigaWeb? Delete Browser data?

you see the config in the developer settings of your browser or also in the log console under ressources...


<img width="1010" alt="image" src="https://github.com/vAmigaWeb/vAmigaWeb/assets/17108995/8d87ebe6-a874-41c8-93c5-e6d2d7855a6d">



in safari you can delete or change it here

<img width="1226" alt="image" src="https://github.com/vAmigaWeb/vAmigaWeb/assets/17108995/1b507e36-eca6-4442-9d3f-79c50412f6d4">

in chrome 

<img width="1226" alt="image" src="https://github.com/vAmigaWeb/vAmigaWeb/assets/17108995/928c2b33-04c9-4919-b731-7c49b93ccbff">

**Comment by dirkwhoffmann** [comment]
> AROS doesn't have SER: it seems

I don't know much about AROS, but I would be surprised if there was no support for the serial port. They use the serial port extensively to dump debug information on startup. 

@wepl How familiar are you with coding at the hardware level? As a workaround, you could simply program the UART directly by writing into the corresponding chipset registers. You would have to configure stuff like the baud rate and implement some interrupt routines that feed the characters to SEROUT.  The HRM contains all the necessary information (I've implemented UART emulation solely on the HRM). 

If I remember correctly, you mentioned the UAE commander at some point. Do you know how they manage the communication between the emulator and the Amiga? Maybe we could steal some ideas.

**Comment by wepl** [comment]
> > AROS doesn't have SER: it seems
> 
> I don't know much about AROS, but I would be surprised if there was no support for the serial port. They use the serial port extensively to dump debug information on startup.

At least when booting only ROM without any files there is no SER: device. Maybe it requires the serial.device on the boot disk. I did not check this (yet).

> @wepl How familiar are you with coding at the hardware level? As a workaround, you could simply program the UART directly by writing into the corresponding chipset registers. You would have to configure stuff like the baud rate and implement some interrupt routines that feed the characters to SEROUT. The HRM contains all the necessary information (I've implemented UART emulation solely on the HRM).

Such a tool would be the best solution I think because it would not require any additional files and configuration.
But it's not done in 10 min and ATM I have no time to write this.
I like first to get it working with existing tools.

> If I remember correctly, you mentioned the UAE commander at some point. Do you know how they manage the communication between the emulator and the Amiga? Maybe we could steal some ideas.

It's memory mapped AFAIK. You write something to specific addresses and the emulator reacts on these accesses.
There is also a logging facility in a similar way.
There is uae-configuration which seems newer that uaectrl. The source seems not to be in the WinUAE github repo.

**Comment by dirkwhoffmann** [comment]
> It's memory mapped AFAIK. You write something to specific addresses and the emulator reacts on these accesses.

The ser: approach appealed to me because it has a clean semantics. In general, I don't like all these special solutions that were only invented for a specific task. On the other hand, let's not overcomplicate things. 

What about this approach? Let's pick an unused chipset register and use it to communicate with RetroShell. Implementation is super easy, as no synchronization is required. The emulator's memory handler (chipset register section) only needs to feed the written value into RetroShell, just as the current serial port implementation does when the RetroShell Commander is plugged in. Optionally, another register could be used to write a symbol to the debug console (in the Web app). 

@mithrendal: Is it OK for you to implement this solution only in the WebApp? I don't see an immediate application for the Desktop app and would like to keep it out for the moment because it is so specialized. (The Desktop app already has RetroShell Commander, which also fits nicely into the existing GUI.)

**Comment by mithrendal** [comment]
@dirkwhoffmann you mean we pick an unused register and the AmigaDOS program moves one byte after the other into it with a closing newline ? When newline is detected the whole char string will be fed into retroshell ? Ok… sounds ok for me.

@wepl Is it something that would be of any value for you ?

**Comment by dirkwhoffmann** [comment]
> you mean we pick an unused register and the AmigaDOS program moves one byte after the other into it with a closing newline ? When newline is detected the whole char string will be fed into retroshell ? Ok… sounds ok for me.

You don't have to wait for a newline character. Whenever the Amiga writes into the register, translate the written value into a key press in RetroShell. Simply translate `\n` to the RetroShell RETURN key.

If you want to implement it more flexibly,  require a prefix to be sent. E.g., if `rsh:xyz` is written into the register, the Amiga starts to feed RetroShell after the `:`. If `js:xyz` is written, you could execute `xyz` in JavaScript etc.

**Comment by wepl** [comment]
> > It's memory mapped AFAIK. You write something to specific addresses and the emulator reacts on these accesses.
> 
> The ser: approach appealed to me because it has a clean semantics. In general, I don't like all these special solutions that were only invented for a specific task. On the other hand, let's not overcomplicate things.

I like the SER: idea. The problem is that using this interface requires some infrastructure on the Amiga side to work:
- a correcty setup devs/system-configuration (no handshake)
- l/port-handler
- devs/serial.device
- kickstart 3.x (maybe also 2.0 works)

As I cannot provide CBM binaries it must work AROS. Until now I have not got it working on AROS.
A solution would be a binary which uses the hw directly without serial.device (as you already proposed). There are some sources on aminet for serial.device replacements. These could probably be used to build such a binary. Still this requires some effort.

> What about this approach? Let's pick an unused chipset register and use it to communicate with RetroShell. Implementation is super easy, as no synchronization is required. The emulator's memory handler (chipset register section) only needs to feed the written value into RetroShell, just as the current serial port implementation does when the RetroShell Commander is plugged in. Optionally, another register could be used to write a symbol to the debug console (in the Web app).
> 
> @mithrendal: Is it OK for you to implement this solution only in the WebApp? I don't see an immediate application for the Desktop app and would like to keep it out for the moment because it is so specialized. (The Desktop app already has RetroShell Commander, which also fits nicely into the existing GUI.)
>
> @dirkwhoffmann you mean we pick an unused register and the AmigaDOS program moves one byte after the other into it with a closing newline ? When newline is detected the whole char string will be fed into retroshell ? Ok… sounds ok for me.
> 
> @wepl Is it something that would be of any value for you ?

For this interface it would be easy to write a program which writes an argument string.

**Reply by mithrendal** [reply]
Regarding the unused custom reg… 
1. which address should we use ?
2. does it have some strange side effects when poking to that register on a real machine ?

**Reply by dirkwhoffmann** [reply]
1. There are several unused register locations at the end of the [custom register area](http://amiga-dev.wikidot.com/information:hardware), e.g., `DFF1F0`.

<img width="760" alt="Bildschirmfoto 2024-03-12 um 23 04 42" src="https://github.com/vAmigaWeb/vAmigaWeb/assets/12561945/b67170cb-bf54-4d2b-b5a9-1e889894499f">

Note that the last address (`DFF1FE`) is used by many programs. It's the official NO-OP register used in Copper instructions that are supposed to do nothing. 
 
2. There are no side-effects when writing into an unused register.

**Comment by mithrendal** [comment]
<img width="1179" alt="image" src="https://github.com/vAmigaWeb/vAmigaWeb/assets/17108995/7e36032b-da09-4811-83ff-17126c2977db">


what about this solution ?

```javascript
hd_mount_list=['UnitA_FA18Interceptor.hdf'];
on_ready_to_run= ()=>
{
  wasm_configure('WARP_MODE','WARP_ALWAYS');
}
on_hdr_step= (drive_number, cylinder) =>
{
  if(drive_number==0 && cylinder==6)
  {
    wasm_configure('WARP_MODE','WARP_NEVER');
  }
}
```

would be very easy to implement on both sides and maybe would do the job sufficiently?

**Reply by wepl** [reply]
I see no way to control an access to a cylinder on the hdd. The placement of files on disk may change when I rebuild the HDF for any reason (e.g. updates). I also need something which works the same way for all installs. The size of the HDF is different for each install. This sounds a bit complicated to use.

**Reply by mithrendal** [reply]
What about this ?
```javascript
hd_mount_list=['UnitA_FA18Interceptor.hdf'];
on_ready_to_run= ()=>
{
  wasm_configure('WARP_MODE','WARP_ALWAYS');
}
on_hdr_step= (drive_number, cylinder) =>
{
  on_power_led_dim = () => wasm_configure('WARP_MODE','WARP_NEVER');
}
```

When the stepper motor of the hd drive moves the first time, it will install a hook on Amiga power led dim event and that triggers the warp_never

In startup sequence you would add a command which dims the Amiga power led, which would stop warp

**Comment by dirkwhoffmann** [comment]
@wepl:  If you favor the chipset register approach, let me know. I can integrate it into the core emulator if there is a demand for it. I.e., it could work like this:

- Writing into `DFF1EC`: 
  The written value is passed to RetroShell as a key-press event.

- Reading from `DFF1EC`:
 RetroShell records its output in a ring buffer. Reading from this memory location returns a character from the buffer, or -1 if it is empty (pretty much like `getc`).

**Comment by wepl** [comment]
> When the stepper motor of the hd drive moves the first time, it will install a hook on Amiga power led dim event and that triggers the warp_never
> 
> In startup sequence you would add a command which dims the Amiga power led, which would stop warp

This would be very easy to handle. Turning warp off via led/audio-filter should be a one shot action as the demo/game/... may also switch this on/off.

> @wepl: If you favor the chipset register approach, let me know. I can integrate it into the core emulator if there is a demand for it. I.e., it could work like this:
> 
> * Writing into `DFF1EC`:
>   The written value is passed to RetroShell as a key-press event.
> * Reading from `DFF1EC`:
>   RetroShell records its output in a ring buffer. Reading from this memory location returns a character from the buffer, or -1 if it is empty (pretty much like `getc`).

This would be more flexible and could probably used for other things too. If the led/audio-filter is working this would be enough for me.

**Reply by mithrendal** [reply]
vAmigaWeb  understands now the two new hooks `on_hdr_step` and  `on_power_led_dim` 😎. Again I did not increment the version number so if no private browseing you have to reinstall the web app. (when we are finally done with this I will increment it so everyone out there will get that version.)

to just try out whether everything is up and running fine ... try this

```javascript
on_hdr_step=()=>{
  alert("hdr stepped for the first time, now install led dim hook");

  on_power_led_dim=()=>{
    alert("power led was dimmed"); 
    on_power_led_dim=()=>{} //after it dimmed remove hook  
  }
  on_hdr_step=()=>{};  // after install of led hook remove the hdr_step hook
}
```

short form of `warp off on power led dim after first hdr step` is this
```javascript
on_hdr_step=()=>on_power_led_dim()=>wasm_configure('WARP_MODE','WARP_NEVER');
```

of course this calls `warp_mode=warp_never` whenever the led dims and it also installs the on led dim hook whenever hdr steps🙄... should not be a problem in terms of performance

a cleaner (but longer) form that removes the hooks after job is done would be this

```javascript
on_hdr_step= (drive_number, cylinder) =>
{
  on_power_led_dim =()=>
  {
    wasm_configure('WARP_MODE','WARP_NEVER'); 
    on_power_led_dim =()=>{}; 
  }
  on_hdr_step=()=>{};
}
```


EDIT:
I looked up how to dim, and found these addresses in vAmiga CIA Code… I bet these asm instructions should do the job… to be tested 
```asm
bset.b #2, $BFE001   //dims power led

bclr.b #2, $BFE001   //strong light on power led 
```

**Reply by wepl** [reply]
Thanks, it doesn't seem to react on led/filter.
See https://whdload.de/ctros/UnitA_FA18InterceptorX.html
I used https://aminet.net/package/util/cli/AFSwitch
afs d

**Reply by mithrendal** [reply]
afs d works fine 

faulty code was ... `on_led_dim` sorry 🙈 ... at first I just named it so but then yesterday I renamed it to `on_power_led_dim`

just rename the hook to `on_power_led_dim` and it should work. I will correct the above example code 🙄


EDIT:
I edited and corrected your html source code inside chromes dev tools locally in my browser as a test ... 

<img width="790" alt="image" src="https://github.com/vAmigaWeb/vAmigaWeb/assets/17108995/9ef948e1-9b59-4986-88ad-7b40af64ee1b">

and it looked promising

**Reply by mithrendal** [reply]
in meantime I did read the manual for the afs command ...

```
afswitch

This is how it works... On a command line you can use one of these
arguments:

D ........ Disable audio filter/LED
E ........ Enable audio filter/LED
S ........ simply Switch the current state of audio filter/LED
H or ? ... small help text
```

I would propose instead of `afs d`, you better call 

```
afs s
afs s
```

i.e. `afs s` two times 

that way you would restore state of the power led before we dim

**Comment by wepl** [comment]
Thanks!
That is working great now. I have rebuild all the ctros.
There are also some demos setup now:
Algotech_TellItToMyHeart.zip
Bastards_TechnoTranceIII.zip
CrionicsSilents_Hardwired.zip
Mirage_Prime90.zip
Sanity_Roots2.zip
StatusOK_Terminator.zip
Underground_Synthesis.zip
WFMH_SiliconGraphics2.zip
WFMH_Software.zip

**Reply by mithrendal** [reply]
<img width="1057" alt="image" src="https://github.com/vAmigaWeb/vAmigaWeb/assets/17108995/45f8a07c-0078-427c-8682-86043e965283">

just watched StatusOK_Terminator ... the beginning takes so long time ... before it really starts 😎. Watched it until the very end... but I needed warp buttons because I was so impatient 🙄... see above I added two of them ... the best would be that the player would get buttons like on netflix 
something like  
 `<< 10s` `10s >>`

**Reply by wepl** [reply]
totally agree 😎

**Comment by wepl** [comment]
I have switched to my own 'nofilter' program because afs hat returned with rc=5.
It's called from whdload via ExecuteStartup in s:whdload.prefs

**Comment by mithrendal** [comment]
Added a warp button to the player and also removed that folder icon which was non functional when there is only one file in the zip-archive … 

<img width="745" alt="image" src="https://github.com/user-attachments/assets/9f13637e-f968-4f21-a7d2-46fadb22dcb9">

When warp is active e.g. on startup the button visually adapts to it so you can see from it that it is warping.
